private-key-compromise/src/agent.ts

```diff
@@ -50,58 +67,111 @@ export const provideHandleTransaction =
     networkManager: NetworkManager<NetworkData>,
     balanceFetcher: BalanceFetcher,
     contractFetcher: ContractFetcher,
-    dataFetcher: DataFetcher
+    dataFetcher: DataFetcher,
+    persistenceHelper: PersistenceHelper,
+    databaseKeys: { transfersKey: string; alertedAddressesKey: string }
   ) =>
   async (txEvent: TransactionEvent) => {
     const findings: Finding[] = [];
 
-    if (txEvent.blockNumber != lastBlock) {
-      lastBlock = txEvent.blockNumber;
-      console.log(`----Transactions processed in block ${txEvent.blockNumber - 1}: ${transactionsProcessed}----`);
-      transactionsProcessed = 0;
-    }
-    transactionsProcessed += 1;
-
-    const transferEvents = txEvent.filterLog(ERC20_TRANSFER_EVENT);
-
     const {
       hash,
       from,
       to,
       transaction: { value, data },
+      timestamp,
+      blockNumber,
     } = txEvent;
 
+    // At the beginning of the block
+    if (blockNumber != lastBlock) {
+      const et = new Date().getTime();
+      console.log(`Block processing time: ${et - st} ms`);
+      const loadedTransferObj = await persistenceHelper.load(databaseKeys.transfersKey.concat("-", chainId));
+      alertedAddresses = await persistenceHelper.load(databaseKeys.alertedAddressesKey.concat("-", chainId));
+
+      transferObj = deepMerge(transferObj, loadedTransferObj);
+
+      // Remove alerted addresses from the object to be persisted into db
+      if (alertedAddresses.length) {
+        for (const el of alertedAddresses) {
+          delete transferObj[el.address];
+        }
+      }
+
+      alertedAddresses = alertedAddresses.filter((address) => timestamp - address.timestamp < TIME_PERIOD);
+
+      lastBlock = blockNumber;
+      console.log(`-----Transactions processed in block ${blockNumber - 7}: ${transactionsProcessed}-----`);
```

Comment:
```suggestion
      console.log(`-----Transactions processed in block ${blockNumber - 7}: ${transactionsProcessed}-----`);
```
Small note: This `${blockNumber - 7}` depends on the number of shards, so as you're using different sharding for each chain you can either use a generic "last block processed" or make the "- 7" dependent on the chain